Move CI script to the rally repository

hub-cap opened the pull request:
This script was previously in a repository that did not allow the rally
team to quickly iterate on, and instead relied on a team that is not
responsible for rally related functionality to approve changes
therein. This commit moves the script into the rally repository so that
it can be iterated on by the rally team.

dliappis commented on .ci/build.sh:
```diff
@@ -0,0 +1,32 @@
+
+# this will only be done if the build number variable is present
+if [ ! -z "$BUILD_NUMBER" ] && [ -d ".rally" ]; then
+  tar -cvjf .rally/$BUILD_NUMBER.tar.bzip $( find .rally/ -name "*.log" )
```
You are assuming here (and in the line above) that the CWD is also where [RALLY_HOME](https://github.com/elastic/rally/commit/b4a1211bb374ac8d0a25e75d45adbfb7d12d4353) is right?

Shouldn't `RALLY_HOME` get honored, if set instead of making this assumption?

hub-cap replied:
much better idea here

dliappis commented on .ci/build.sh:
```diff
@@ -0,0 +1,32 @@
+make it
+
+# this will only be done if the build number variable is present
+if [ ! -z "$BUILD_NUMBER" ] && [ -d ".rally" ]; then
```
It think it'd be valueable to be stricter about which shell we use in the shebang and specify `#!/usr/bin/env bash` in line 1 and avoid surprises with `dash` or plain `sh`.

If that's done we could perhaps use the [bash conditionals](https://www.gnu.org/software/bash/manual/bash.html#Conditional-Constructs) `[[` / `]]` for these tests.

hub-cap replied:
wow, what a rookie mistake to forget the shebang.

hub-cap commented:
ok I have updated the script. The general idea here is that an external caller can execute the build and archive steps independent of each other, so if build fails the caller can still execute archive. this is why it is now 2 functions and the caller must specify the function they wish to run. I have also addressed reviewer comments.

dliappis commented on .ci/build.sh:
```diff
@@ -0,0 +1,57 @@
+  if [[ -d ${RALLY_DIR} ]]; then
+    find ${RALLY_DIR} -name "*.log" -printf "%P\\0" | tar -cvjf ${RALLY_DIR}/${BUILD_NUMBER}.tar.bz2 -C ${RALLY_DIR} --transform "s,^,ci-${BUILD_NUMBER}/," --null -T -
+  else
+    echo "Rally directory not present, this should not happen"
```
Maybe print the actual value of `${RALLY_DIR}` for troubleshooting too?

dliappis commented on .ci/build.sh:
```diff
@@ -0,0 +1,57 @@
+  fi
+}
+
+if declare -f "$1" > /dev/null; then
```
We only need to check whether the function name exists right?

According to the man page, `-f` shows also the content which is not needed. Instead we can use `-F` which only shows the function name:

> The -f option will restrict the display to shell functions.  The -F option inhibits the display of function definitions; only the function name and attributes are printed.

Consider the following example:

```
$ function myf {
> echo "sdfsdf"
> }
```

```
$ declare -f myf
myf ()
{
    echo "sdfsdf"
}
```

```
## function exists
$ declare -F myf
myf
$ echo $?
0
```

```
## function not found
$ declare -F myf2
$ echo $?
1
```

hub-cap replied:
nice. I took this from https://github.com/elastic/rally/blob/master/integration-test.sh#L634 , shall I change it there as well?

dliappis replied:
can be done, but in a separate PR/commit plz
